codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,136 +16,248 @@
+ *
+ * <h3>HTTP Upgrade</h3>
+ *
+ * Server-side HTTP to HTTP/2 upgrade is supported in conjunction with {@link Http2ServerUpgradeCodec}; the necessary
+ * HTTP-to-HTTP/2 conversion is performed automatically.
  */
 @UnstableApi
 public class Http2FrameCodec extends ChannelDuplexHandler {
 
-    private static final Http2FrameLogger HTTP2_FRAME_LOGGER = new Http2FrameLogger(INFO, Http2FrameCodec.class);
+    private static final InternalLogger LOG = InternalLoggerFactory.getInstance(Http2FrameCodec.class);
 
     private final Http2ConnectionHandler http2Handler;
     private final boolean server;
+    private final PropertyKey streamKey;
+
+    // Used to adjust flow control window on channel active. Set to null afterwards.
+    private Integer initialLocalConnectionWindow;
 
     private ChannelHandlerContext ctx;
     private ChannelHandlerContext http2HandlerCtx;
 
+    private Http2Stream2Impl pendingOutboundStreamsTail;
+
+    /** Lock protecting modifications to idle outbound streams. **/
+    private final Object lock = new Object();
+
+    /** Number of buffered streams if the {@link StreamBufferingEncoder} is used. **/
+    private int numBufferedStreams;
+
     /**
-     * Construct a new handler.
-     *
-     * @param server {@code true} this is a server
+     * Create a new handler. Use {@link Http2FrameCodecBuilder}.
      */
-    public Http2FrameCodec(boolean server) {
-        this(server, HTTP2_FRAME_LOGGER);
+    Http2FrameCodec(Http2ConnectionEncoder encoder, Http2ConnectionDecoder decoder, Http2Settings initialSettings,
+                    long gracefulShutdownTimeoutMillis) {
+        decoder.frameListener(new FrameListener());
+        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, initialSettings);
+        http2Handler.connection().addListener(new ConnectionListener());
+        http2Handler.gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
+        server = http2Handler.connection().isServer();
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
     }
 
     /**
-     * Construct a new handler.
+     * Creates a new outbound/local stream.
+     *
+     * <p>The object is added to a list of idle streams, so that in case the stream object is never made active, the
+     * {@link Http2Stream2#closeFuture()} still completes.
+     *
+     * <p>This method may only be called after the handler has been added to a {@link io.netty.channel.ChannelPipeline}.
      *
-     * @param server {@code true} this is a server
+     * <p>This method is thread-safe.
      */
-    public Http2FrameCodec(boolean server, Http2FrameLogger frameLogger) {
-        this(server, new DefaultHttp2FrameWriter(), frameLogger);
-    }
+    // TODO(buchgr): Discuss: Should this method be thread safe?
+    Http2Stream2 newStream() {
```

Comment:
@ejona86  @Scottmitch @normanmaurer @nmittler @louiscryan 

The `newStream()` method allows one to create a new `Http2Stream2` object (without a stream identifier). This method is thread-safe. The idea is that a user can call it from within or outside the eventloop to obtain the object and then attach it to the first headers frame. The frame codec will then assign a stream identifier to the object. 

```java
final Http2Stream2 newStream = newStream();
ctx.write(new DefaultHttp2HeadersFrame(headers).stream(newStream))
  .addListener(new ChannelFutureListener() {
    @Override
    public void operationComplete(ChannelFuture f) {
       if (f.isSuccess()) {
         // stream is active and newStream.id() has a valid value.
       }
    }
  });
ctx.write(new DefaultHttp2DataFrame(bytebuf).stream(newStream);
```

So far so good. During development I noticed that it's pretty useful for one to be able to perform some action when a stream closes. I initially had a method `void Http2Stream2#onStreamClosed(Runnable task)` in mind, but after a discussion with @ejona86 , he suggested that a better way is to have `ChannelFuture Http2Stream2#closeFuture()`, where one can attach listeners that get executed when a stream closes. Having a `ChannelFuture` is nice, as it feels more like Netty and also makes the threading clear: all listeners are executed in the channel's eventloop.

An implication of this is, that a `ChannelFuture` needs a `Channel`. So `newStream()` may only be called after the handler was added to the pipeline. This might complicate things when calling `newStream()` from outside the eventloop. I think that's fine though, because Netty has the same problem with writing from outside of the eventloop. Netty is just not nice if you don't want to block at all on the client side (not even until a channel is registered).

However, one problem to think about is what to do with `Http2Stream2` objects that are never mapped to an active stream. What "guarantees" should we give for `Http2Stream2#closeFuture()`?

```java
Http2Stream2 newStream = newStream();
newStream.closeFuture().addListener(failureListener);
... 
if (someCondition) {
  ctx.write(new DefaultHttp2HeadersFrame(headers).stream(newStream));
}
```

What if `someCondition` is `false` and `newStream` never reaches the `Http2FrameCodec`. What if an `Exception` is thrown before the write? Should that `failureListener` still be executed eventually? In the current implementation the answer to this question is yes. That is, the frame codec maintains a list of pending streams. Pending streams are streams that never become active. If the frame codec is removed from the pipeline or the channel is closed the `closeFutures` of all pending streams are completed. However, the implementation of pending streams is complex and requires locking (or someone to implement a lock free doubly linked list). All that to handle an edge case that most applications will _never_ encounter, but for which _all_ applications pay a performance penalty.

So an alternative would be, that only after a stream becomes active the `closeFuture` is guaranteed to be completed.

```java
Http2Stream2 newStream = newStream();
// This listener might never be executed
newStream.closeFuture().addListener(someListener);
ctx.write(new DefaultHttp2Headers(header).stream(newStream))
.addListener(new ChannelFutureListener() {
  @Override
  public void operationComplete(ChannelFuture f) {
    if (f.isSuccess()) {
      // From here on it's guaranteed that the closeFuture completes after a stream is closed.
      newStream.closeFuture().addListener(otherListener);
    }
  }
}
```

I think I would prefer the second API, however, I can also see that users might not read the API documentation carefully and write buggy applications. Alternatively, we  could also call the method `Http2Stream2#closeFutureThatWillOnlyCompleteAfterTheStreamBecameActive()` :)

Thoughts?